examples/reporting/parallel_report_download.py

```diff
@@ -0,0 +1,159 @@
+    ad_group_query = """
+        SELECT campaign.id, ad_group.id, metrics.impressions, metrics.clicks
+        FROM ad_group
+        WHERE segments.date DURING LAST_30_DAYS"""
+
+    inputs = _generate_inputs(
+        client, customer_ids, [campaign_query, ad_group_query]
+    )
+    with multiprocessing.Pool(MAX_PROCESSES) as pool:
+        results = pool.starmap(_issue_search_request, inputs)
+        print(results)
+
+
+def _issue_search_request(client, customer_id, query):
+    """Issues a search request using streaming.
+
+    If a 500 error is received, retries until MAX_RETRIES is reached.
+
+    Args:
+        client: an initialized GoogleAdsClient instance.
+        customer_id: a client customer ID str.
+        query: a GAQL query str.
+    """
+    ga_service = client.get_service("GoogleAdsService", version="v6")
+    retry_count = 0
+    # A while loop allows us to retry until we've reached MAX_RETRIES or
+    # have successfully received a response.
+    while True:
+        try:
+            response = ga_service.search_stream(customer_id, query)
+            # Returning a list of GoogleAdsRows will result in a
+            # PicklingError, so instead we put the GoogleAdsRow data
+            # into a list of str results and return that.
+            result_strings = []
+            for batch in response:
+                for row in batch.results:
+                    ad_group_id = (
+                        f"Ad Group ID {row.ad_group.id} in "
+                        if "ad_group.id" in query
+                        else ""
+                    )
+                    result_string = (
+                        f"{ad_group_id}"
+                        f"Campaign ID {row.campaign.id} "
+                        f"had {row.metrics.impressions} impressions "
+                        f"and {row.metrics.clicks} clicks"
+                    )
+                    result_strings.append(result_string)
+            return result_strings
+        except GoogleAdsException as ex:
+            if retry_count < MAX_RETRIES:
+                retry_count += 1
+                time.sleep(retry_count * BACKOFF_FACTOR)
+            else:
+                print(
+                    f"Request for customer ID {customer_id} with request ID "
+                    f"{ex.request_id} failed with status "
+                    f"{ex.error.code().name} after {retry_count + 1} "
+                    "retries and includes the following errors:"
+                )
+                for error in ex.failure.errors:
+                    print(f'\tError with message "{error.message}".')
+                    if error.location:
+                        for (
+                            field_path_element
+                        ) in error.location.field_path_elements:
+                            print(
+                                f"\t\tOn field: {field_path_element.field_name}"
+                            )
+                sys.exit(1)
```

Comment:
You need to show better error handling. In this case the `sys.exit` will cause just the worker process to fail. But if there are multiple concurrent downloads all failing due to retries, there should be a signal to the controller process to cease everything, otherwise it may compound a bad backend problem or potentially lead to rate limiting.

Reply:
What do you think of counting retries globally and handling exceptions with `startmap_async`'s error_callback param? e.g.:
1. edit `pool.starmap` to:
```
with multiprocessing.Pool(MAX_PROCESSES) as pool:
        retries = 0
        results = pool.starmap_async(
            _issue_search_request,
            inputs,
            error_callback=lambda ex: _handle_error(ex, retries))
```
2. re-raise the exception in `_issue_search_request`
3. handle errors with: 
```
def _handle_error(exception, retries):
    if retries < MAX_RETRIES:
        retries += 1
        time.sleep(retries * BACKOFF_FACTOR)
    else:
        print(f'Failed with exception {exception}')
        sys.exit(1)
```
If I keep all else the same, I believe that would mean n (MAX_RETRIES) retries per account/query, and if n separate account/query combos failed, the program controller process would exit. 

We could also maintain separate variables for retry counts within a process vs globally, if these should not be the same. What do you think?